apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts

```diff
@@ -362,39 +440,79 @@ Keep in mind you are not just generating an example; you are generating an actua
     user: openAiUserFromAuthn(authnUserId),
   });
 
-  extractFromCompletion(completion, job);
-  job.data['generation'] = completion.choices[0].message.content;
-  job.data['completion'] = completion;
+  const results = extractFromCompletion(completion, job);
 
-  const html = job?.data?.html || originalHTML;
+  const html = results?.html || originalHTML;
 
-  if (saveInitialErrors) {
-    job.data['initialGenerationErrors'] = [];
-  }
-  job.data['finalGenerationErrors'] = [];
+  let errors: string[] = [];
 
   if (html && typeof html === 'string') {
-    const errors = validateHTML(html, false, !!job?.data?.python);
-    if (saveInitialErrors) {
-      job.data['initialGenerationErrors'] = errors;
+    errors = validateHTML(html, false, !!results?.python);
+  }
+
+  if (userId !== undefined && hasCoursePermissionEdit !== undefined) {
+    await queryAsync(sql.insert_ai_generation_prompt, {
+      question_id: questionId,
+      prompting_user_id: authnUserId,
+      prompt_type: isAutomated ? 'auto_revision' : 'human_revision',
+      user_prompt: revisionPrompt,
+      system_prompt: sysPrompt,
+      response: completion.choices[0].message.content,
+      html: results?.html,
+      python: results?.python,
+      errors,
+      completion,
+      job_sequence_id: jobSequenceId,
+    });
+
+    const files: Record<string, string> = {};
+    if (results?.html) {
+      files['question.html'] = results?.html;
     }
-    job.data['finalGenerationErrors'] = errors;
-    if (errors.length > 0 && numRegens > 0) {
-      const autoRevisionPrompt = `Please fix the following issues: \n${errors.join('\n')}`;
-      job.data['autoRevisionPrompt'] = autoRevisionPrompt;
-      await regenInternal(
-        job,
-        client,
-        authnUserId,
-        originalPrompt,
-        autoRevisionPrompt,
-        html,
-        typeof job?.data?.python === 'string' ? job?.data?.python : undefined,
-        numRegens - 1,
-        false,
-      );
+
+    if (results?.python) {
+      files['server.py'] = results?.python;
+    }
+
+    const client = getCourseFilesClient();
+
+    const result = await client.updateQuestionFiles.mutate({
+      course_id: courseId,
+      user_id: userId,
+      authn_user_id: authnUserId,
+      has_course_permission_edit: hasCoursePermissionEdit,
+      question_id: questionId,
+      files,
+    });
+
+    if (result.status === 'error') {
+      throw Error('Draft mutation failed.');
     }
   }
+
+  job.data.html = html;
+  job.data.python = results.python;
+
+  if (errors.length > 0 && numRegens > 0) {
+    const auto_revisionPrompt = `Please fix the following issues: \n${errors.join('\n')}`;
+    await regenInternal({
+      job,
+      client,
+      authnUserId,
+      originalPrompt,
+      revisionPrompt: auto_revisionPrompt,
+      originalHTML: html,
+      originalPython: typeof job?.data?.python === 'string' ? job?.data?.python : undefined,
+      numRegens: numRegens - 1,
```

Comment:
Nit: I was reading this as "number of regenerations attempted so far", but it seems like this is actually "number of attempts left". Maybe this should be named `remainingAttempts`?

Reply:
changed in d830080b8b22c8440c018379442f76ca019db963.